File: lib/tesla_mate/log/position.ex
defmodule TeslaMate.Log.Position do
  use Ecto.Schema
  import Ecto.Changeset

  alias TeslaMate.Log.{Car, Trip}

  schema "positions" do
    field :date, :utc_datetime
    field :latitude, :float
    field :longitude, :float

    field :speed, :integer
    field :power, :integer
    field :odometer, :float
    field :ideal_battery_range_km, :float
    field :battery_level, :float
    field :outside_temp, :float
    field :altitude, :float

    belongs_to(:car, Car)
    belongs_to(:trip, Trip)
  end

  @doc false
  def changeset(position, attrs) do
    position
    |> cast(attrs, [
      :date,
      :latitude,
      :longitude,
      :speed,
      :power,
      :odometer,
      :ideal_battery_range_km,
      :battery_level,
      :outside_temp,
      :altitude
    ])
    |> validate_required([:car_id, :date, :latitude, :longitude])
    |> foreign_key_constraint(:car_id)
    |> foreign_key_constraint(:trip_id)
  end
end

File: ios/TeslaMateApp/TeslaMateApp/Models/Position.swift
import Foundation
import CoreLocation

struct Position: Codable, Identifiable {
    let id: Int
    let date: String?
    let latitude: Double?
    let longitude: Double?
    let elevation: Int?
    let speed: Int?
    let power: Int?
    let odometer: Double?
    let batteryLevel: Int?
    let usableBatteryLevel: Int?
    let idealBatteryRangeKm: Double?
    let estBatteryRangeKm: Double?
    let ratedBatteryRangeKm: Double?
    let outsideTemp: Double?
    let insideTemp: Double?
    let fanStatus: Int?
    let isClimateOn: Bool?

    var coordinate: CLLocationCoordinate2D? {
        guard let latitude, let longitude else { return nil }
        return CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }
}
